README
==================================
Axel's Tiling Window Manager Setup
==================================

History, Background and Reasoning
---------------------------------

While being a heavy FVWM user for nearly 15 years, I now run tiling
window managers on all my "desktops" and laptops, ranging from a first
generation ASUS EeePC 701 4G netbook over a Sun UltraSparc 10 to a
full-fledged multihead workstation from Dalco at work.

My ASUS EeePC 701 running Debian GNU/Linux Sid was the first box where
I did not use FVWM anymore, because with a resolution of 800x480 on a
7 inch screen, you a) want to waste as few pixels as possible by title
bars, borders, etc. and b) most of the time anything else than
fullscreen windows doesn't make sense, moving windows around with the
mouse makes even less sense.

So I've chosen ratpoison for the EeePC (and still use it on that box)
since it makes windows fullscreen by default and the keybindings are
nearly identical to GNU Screen and therefore easy to learn
respectively I didn't have to learn anything to use it.

After more and more fine tuning involving the xmobar text status bar,
filled with data by i3status, I started this git repository to track
my own changes, to share the setup with some other of my boxes (like
the UltraSparc or my bed-side terminal, an older ThinkPad A31).

Getting used with a fullscreen and tiling setup on the EeePC I more
and more wanted something for my everyday ThinkPad T61, too. But while
ratpoison is really perfect for the small EeePC screen, it proved too
clumsy for more complex window arrangements and multiple virtual
desktops.

So I looked through the other tiling window managers in Debian, trying
out i3, spectrwm (formerly scrotwm), wmii and awesome and some ion
successors (tritium and anion). I finally stuck with awesome, first
the 2.x version from Debian 5.0 Lenny, now the 3.x version from Debian
6.0 Squeeze.

In general I liked the idea of using the -- on Linux mostly unused --
Windows key as window manager meta key. I even configured my ratpoison
to use that in addition to the original Ctrl-T prefix.

You will also find some configuration for i3 (version 4 and upwards)
in the repo as I also experiment with i3's tree-based layout to see if
I can workaround some of awesome's shortcomings by using i3 instead.


Repository Name and URLs
------------------------

So this repository is no more a ratpoison only setup. But since I
neither want to change the repository name nor any URL I decided that
I stick with having "ratpoison" in the name. The amount of "rodent"
usage you need with this setup hasn't changed anyway, so without the
relation to the window manager of this name, the name is still fitting
(as it fits to the window manager of that name :-).


Installation
------------

These files are needed to get the setup running. The setup currently
runs on Debian Sid and Squeeze as well as Ubuntu Natty. It probably
also runs on older Ubuntu releases.

Since most of the configuration files included are not expected to
reside in .ratpoison, some symlinks are necessary where config file
paths cannot be set via command line options or where it would be to
tedious to always type them:

~/.gitconfig 				   → gitconfig
~/.xsession 				   → xsession
~/.screenrc 				   → screenrc
~/.colordiffrc 				   → colordiffrc
~/bin/xrandr-vga-off.sh			   → bin/xrandr-vga-off.sh
~/bin/terms				   → bin/terms

Sources available via http://git.noone.org/?p=ratpoison.git,
http://gitorious.org/abe/ratpoison-desktop, and
http://github.com/xtaran/ratpoison-desktop.

Needs at least the following Debian packages (besides essential
packages) to be installed:

* busybox-syslogd (for logread)
* conkeror or some other web browser
* emacs (or emacsen, for emacs and emacsclient)
* gnome-keyring (for gnome-settings-daemon)
* hsetroot
* i3status >= 2.2
* libfile-slurp-perl
* libfile-temp-perl
* lsb-release
* ratmenu >= 2.3.20
* ratpoison or awesome + awesome-extra
* wget (needed by iplet)
* x11-utils (for xmessage) or gxmessage
* x11-xserver-utils (for xmodmap, xrdb, xrandr and xsetroot)
* xmobar
* xscreensaver or xtrlock
* xterm and/or rxvt-unicode
* For Focus-Follows-Mouse:
  * for awesome: xdotool
  * for ratpoison: xdotool or nawm (no more in Debian)

Fonts used for xmobar:

* ttf-mplus | fonts-mplus
* ttf-droid | fonts-droid
* xfonts-terminus

Used if available but except the system tray stuff recommended anyway:

* autocutsel
* inotail
* keynav >= 0.20101014.3067
* redshift
* unclutter
* x11-xkb-utils (for setxkbmap)
* xrootconsole
* xserver-xorg-input-synaptics (for synclient+syndaemon)
* yeahconsole
* System tray applets for use with awesome:
  * nm-applet
  * wicd-gtk
  * update-notifier
  * gtk-redshift
  * fdpowermon
  * radiotray
  * shutter
  * yarssr
  * qasmixer | volumeicon-alsa

Linux kernel modules which may be used by some features of xmobar, but
do not seem to be loaded automatically (write them into /etc/modules):

* acpi_cpufreq
* coretemp

Only used in non-necessary scripts or keybindings:

* dwm-tools >= 31-1 or suckless-tools (for dmenu, tabbed and wmname)
* openssh-client (for ssh-add and ssh-agent)
* screen
* scrot
* transset-df
* xclip
* colordiff

Only in commented code (i.e currently not used):

* Alternative window and session managers:
  * i3
  * flwm
  * fvwm
  * lxsession
  * spectrwm (formerly scrotwm)
  * stumpwm
  * tritium (or not yet packaged: anion3 and notion)
* loco, ccze, lwatch, or colortail
* root-tail
* xcompmgr


debian/changelog
abe-desktop (4) unstable; urgency=low

  * Suggest radiotray, yarssr and qasmixer | volumeicon-alsa in
    abe-desktop-awesome.
  * Suggest audacious and audacious-plugins in abe-desktop-common
  * Alternatively depend on emacsen, not only emacs.

 -- Axel Beckert <abe@debian.org>  Wed, 18 Jul 2012 10:21:14 +0200

abe-desktop (3) unstable; urgency=low

  * nm-applet is no separate package but part of network-manager-gnome
  * Loosen dependencies on *-droid and i3status >= 2.2 as those are not
    available in squeeze (and partially not even in squeeze backports)

 -- Axel Beckert <abe@debian.org>  Tue, 27 Dec 2011 19:17:51 +0100

abe-desktop (2) unstable; urgency=low

  * Allow other syslog daemons besides busybox-syslogd as Ubuntu's meta
    packages hardcode rsyslog.

 -- Axel Beckert <abe@debian.org>  Sat, 24 Dec 2011 15:49:58 +0100

abe-desktop (1) unstable; urgency=low

  * Make mplus fonts optional as they're not in use everywhere.
  * Get version from changelog instead of control file.

 -- Axel Beckert <abe@deuxchevaux.org>  Sat, 24 Dec 2011 15:33:16 +0100

abe-desktop (0) unstable; urgency=low

  * Initial package

 -- Axel Beckert <abe@deuxchevaux.org>  Sat, 24 Dec 2011 14:32:20 +0100



awesome/rc.lua
-- Standard awesome library
require("awful")
require("awful.autofocus")
require("awful.rules")
require("awful.remote")
-- Theme handling library
require("beautiful")
-- Notification library
require("naughty")

-- Load Debian menu entries
require("debian.menu")

-- Load obvious stuff
require("obvious.battery")
require("obvious.volume_alsa")

-- {{{ Variable definitions
-- Themes define colours, icons, and wallpapers
beautiful.init((os.getenv("RP") or (os.getenv("HOME") .."/.config")) .. "/awesome/theme.lua")
-- awful.util.spawn("awsetbg -l", false, s)

-- Only set bottom padding for screen 1
awful.screen.padding( screen[1], { bottom = 16 } )

-- This is used later as the default terminal and editor to run.
terminal = "x-terminal-emulator"
editor = "emacsclient -a emacs"
editor_cmd = "emacsclient -a emacs"
screen_cmd = terminal .. " -e screen -R"
browser = "sensible-browser"
xlock = "xscreensaver-command -activate || xtrlock"
amixer = "amixer"
ffm = (os.getenv("RP") or (os.getenv("HOME") .."/.ratpoison"))  .. "/bin/focus-follows-mouse-on-tag-change.sh"

-- Determine hostname
local io = { popen = io.popen }
local f = io.popen("hostname")
local hostname = f:read("*all")
f:close()
hostname = string.gsub(hostname, '[\n\r]+', '')

-- Default modkey.
-- Usually, Mod4 is the key with a logo between Control and Alt.
-- If you do not like this or do not have such a key,
-- I suggest you to remap Mod4 to another key using xmodmap or other tools.
-- However, you can use another modifier like Mod1, but it may interact with others.
modkey = "Mod4"

-- Table of layouts to cover with awful.layout.inc, order matters.
layouts =
{
    awful.layout.suit.tile,            --  1
    awful.layout.suit.tile.bottom,     --  2
    awful.layout.suit.tile.left,       --  3
    awful.layout.suit.tile.top,        --  4
    awful.layout.suit.fair,            --  5
    awful.layout.suit.fair.horizontal, --  6
    awful.layout.suit.spiral,          --  7
    awful.layout.suit.spiral.dwindle,  --  8
    awful.layout.suit.magnifier,       --  9
    awful.layout.suit.max.fullscreen,  -- 10
    awful.layout.suit.max,             -- 11
    awful.layout.suit.floating         -- 12
}
-- }}}

-- {{{ Tags
-- Define a tag table which hold all screen tags.
tags = {
   settings = {
      {
	 names  = { 1, 2, 3, 4, 5, 6, 7, 8, 9,
		    10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22 },
	 layout = { layouts[1], layouts[1], layouts[1], layouts[1], layouts[1],
		    layouts[1], layouts[1], layouts[1], layouts[1], layouts[1],
		    layouts[1], layouts[1], layouts[1], layouts[1], layouts[1],
		    layouts[1], layouts[1], layouts[1], layouts[1], layouts[1],
		    layouts[1], layouts[1] }
}}}

for s = 1, screen.count() do
  -- Fill tag table with potentially host specific stuff
  if hostname == "snidget" and ( s == 2 or s == 3 ) then
     tags.settings[s] = {
	names = { 1, 2, 3, 4, 5, 6, 7, 8, 9 },
	layout = { layouts[2], layouts[2], layouts[2], layouts[2], layouts[2],
		   layouts[2], layouts[2], layouts[2], layouts[2] }
     }
  else
     if s ~= 1 then
	tags.settings[s] = tags.settings[1]
     end
  end
  -- Each screen has its own tag table.
  tags[s] = awful.tag(tags.settings[s].names, s, tags.settings[s].layout)
  -- On snidget preset specific tag configurations
  if hostname == "snidget" then
     awful.tag.setnmaster(2, tags[1][1])
     awful.tag.setncol(2,    tags[1][1])
     if s == 2 or s == 3 then
	awful.tag.setmwfact(.85, tags[s][1])
     end
     if s == 4 then
	awful.tag.setmwfact(.7, tags[s][1])
     end
  end
end
-- }}}

-- {{{ Menu
-- Create a laucher widget and a main menu
myawesomemenu = {
   { "manual", terminal .. " -e man awesome" },
   { "edit config", editor_cmd .. " " .. awful.util.getdir("config") .. "/rc.lua" },
   { "restart", awesome.restart },
   { "quit", awesome.quit }
}

mymainmenu = awful.menu({ items = { { "awesome", myawesomemenu, beautiful.awesome_icon },
                                    { "Debian", debian.menu.Debian_menu.Debian },
                                    { "open terminal", terminal }
                                  }
                        })

mylauncher = awful.widget.launcher({ image = image(beautiful.awesome_icon),
                                     menu = mymainmenu })
-- }}}

-- {{{ Wibox
-- Create a textclock widget
mytextclock = awful.widget.textclock({ align = "right" })

-- Create a systray for each screen
mysystray = {}
for s = 1, screen.count() do
   mysystray[s] = widget({ type = "systray" })
end

-- Create a wibox for each screen and add it
mywibox = {}
mypromptbox = {}
mylayoutbox = {}
mytaglist = {}
mytaglist.buttons = awful.util.table.join(
                    awful.button({ }, 1, awful.tag.viewonly),
                    awful.button({ modkey }, 1, awful.client.movetotag),
                    awful.button({ }, 3, awful.tag.viewtoggle),
                    awful.button({ modkey }, 3, awful.client.toggletag),
                    awful.button({ }, 4, awful.tag.viewnext),
                    awful.button({ }, 5, awful.tag.viewprev)
                    )
mytasklist = {}
mytasklist.buttons = awful.util.table.join(
                     awful.button({ }, 1, function (c)
                                              if not c:isvisible() then
                                                  awful.tag.viewonly(c:tags()[1])
                                              end
                                              client.focus = c
                                              c:raise()
                                          end),
                     awful.button({ }, 3, function ()
                                              if instance then
                                                  instance:hide()
                                                  instance = nil
                                              else
                                                  instance = awful.menu.clients({ width=250 })
                                              end
                                          end),
                     awful.button({ }, 4, function ()
                                              awful.client.focus.byidx(1)
                                              if client.focus then client.focus:raise() end
                                          end),
                     awful.button({ }, 5, function ()
                                              awful.client.focus.byidx(-1)
                                              if client.focus then client.focus:raise() end
                                          end))

for s = 1, screen.count() do
    -- Create a promptbox for each screen
    mypromptbox[s] = awful.widget.prompt({ layout = awful.widget.layout.horizontal.leftright })
    -- Create an imagebox widget which will contains an icon indicating which layout we're using.
    -- We need one layoutbox per screen.
    mylayoutbox[s] = awful.widget.layoutbox(s)
    mylayoutbox[s]:buttons(awful.util.table.join(
                           awful.button({ }, 1, function () awful.layout.inc(layouts, 1) end),
                           awful.button({ }, 3, function () awful.layout.inc(layouts, -1) end),
                           awful.button({ }, 4, function () awful.layout.inc(layouts, 1) end),
                           awful.button({ }, 5, function () awful.layout.inc(layouts, -1) end)))
    -- Create a taglist widget
    mytaglist[s] = awful.widget.taglist(s, awful.widget.taglist.label.all, mytaglist.buttons)

    -- Create a tasklist widget
    mytasklist[s] = awful.widget.tasklist(function(c)
                                              return awful.widget.tasklist.label.currenttags(c, s)
                                          end, mytasklist.buttons)

    -- Create the wibox
    mywibox[s] = awful.wibox({ position = "top", screen = s })
    -- Add widgets to the wibox - order matters
    mywibox[s].widgets = {
        {
            mylauncher,
            mytaglist[s],
            mypromptbox[s],
            layout = awful.widget.layout.horizontal.leftright
        },
        mylayoutbox[s],
	obvious.battery(),
        mytextclock,
        mysystray[s],
        mytasklist[s],
        layout = awful.widget.layout.horizontal.rightleft
    }
end
-- }}}

-- {{{ Mouse bindings
root.buttons(awful.util.table.join(
    awful.button({ }, 3, function () mymainmenu:toggle() end),
    awful.button({ }, 4, awful.tag.viewnext),
    awful.button({ }, 5, awful.tag.viewprev)
))
-- }}}

-- {{{ Key bindings
globalkeys = awful.util.table.join(
    awful.key({ modkey, "Control" }, "apostrophe", awful.tag.viewnone),
    awful.key({ modkey,           }, "Left",
	function ()
	    awful.tag.viewprev()
	    awful.util.spawn(ffm)
	end),
    awful.key({ modkey,           }, "Right",
	function ()
	    awful.tag.viewnext()
	    awful.util.spawn(ffm)
	end),
    awful.key({ modkey,           }, "Escape", awful.tag.history.restore),

    awful.key({ modkey,           }, "j",
        function ()
            awful.client.focus.byidx( 1)
            if client.focus then client.focus:raise() end
        end),
    awful.key({ modkey,           }, "k",
        function ()
            awful.client.focus.byidx(-1)
            if client.focus then client.focus:raise() end
        end),
    awful.key({ modkey,           }, "w", function () mymainmenu:show({keygrabber=false}) end),

    -- Layout manipulation
    awful.key({ modkey, "Shift"   }, "j", function () awful.client.swap.byidx(  1)    end),
    awful.key({ modkey, "Shift"   }, "k", function () awful.client.swap.byidx( -1)    end),
    awful.key({ modkey, "Control" }, "j", function () awful.screen.focus_relative( 1) end),
    awful.key({ modkey, "Control" }, "k", function () awful.screen.focus_relative(-1) end),
    awful.key({ modkey,           }, "u", awful.client.urgent.jumpto),
    awful.key({ modkey,           }, "Tab",
        function ()
            awful.client.focus.history.previous()
            if client.focus then
                client.focus:raise()
            end
        end),

    -- Standard program
    awful.key({ modkey,           }, "Return", function () awful.util.spawn(terminal) end),
    awful.key({ modkey, "Control" }, "r", awesome.restart),
    awful.key({ modkey, "Shift"   }, "q", awesome.quit),

    awful.key({ modkey,           }, "l",     function () awful.tag.incmwfact( 0.05)    end),
    awful.key({ modkey,           }, "h",     function () awful.tag.incmwfact(-0.05)    end),
    awful.key({ modkey, "Shift"   }, "h",     function () awful.tag.incnmaster( 1)      end),
    awful.key({ modkey, "Shift"   }, "l",     function () awful.tag.incnmaster(-1)      end),
    awful.key({ modkey, "Control" }, "h",     function () awful.tag.incncol( 1)         end),
    awful.key({ modkey, "Control" }, "l",     function () awful.tag.incncol(-1)         end),
    awful.key({ modkey,           }, "space", function () awful.layout.inc(layouts,  1) end),
    awful.key({ modkey, "Shift"   }, "space", function () awful.layout.inc(layouts, -1) end),

    -- Prompt
    awful.key({ modkey },            "r",     function () mypromptbox[mouse.screen]:run() end),

    awful.key({ modkey }, "x",
              function ()
                  awful.prompt.run({ prompt = "Run Lua code: " },
                  mypromptbox[mouse.screen].widget,
                  awful.util.eval, nil,
                  awful.util.getdir("cache") .. "/history_eval")
              end)
)

clientkeys = awful.util.table.join(
    awful.key({ modkey,           }, "f",      function (c) c.fullscreen = not c.fullscreen  end),
    awful.key({ modkey, "Shift"   }, "c",      function (c) c:kill()                         end),
    awful.key({ modkey, "Control" }, "space",  awful.client.floating.toggle                     ),
    awful.key({ modkey, "Control" }, "Return", function (c) c:swap(awful.client.getmaster()) end),
    awful.key({ modkey,           }, "o",      awful.client.movetoscreen                        ),
    awful.key({ modkey, "Shift"   }, "r",      function (c) c:redraw()                       end),
    awful.key({ modkey,           }, "t",      function (c) c.ontop = not c.ontop            end),
    awful.key({ modkey,           }, "n",      function (c) c.minimized = not c.minimized    end),
    awful.key({ modkey,           }, "m",
        function (c)
            c.maximized_horizontal = not c.maximized_horizontal
            c.maximized_vertical   = not c.maximized_vertical
        end)
)

-- Compute the maximum number of digit we need, limited to 9
keynumber = 0
for s = 1, screen.count() do
   keynumber = math.min(10, math.max(#tags[s], keynumber));
end

-- Bind all key numbers to tags.
-- Be careful: we use keycodes to make it works on any keyboard layout.
-- This should map on the top row of your keyboard, usually 1 to 9.
for i = 1, keynumber do
    globalkeys = awful.util.table.join(globalkeys,
        awful.key({ modkey }, "#" .. i + 9,
                  function ()
                        local screen = mouse.screen
                        if tags[screen][i] then
                            awful.tag.viewonly(tags[screen][i])
			    awful.util.spawn(ffm)
                        end
                  end),
        awful.key({ modkey, "Control" }, "#" .. i + 9,
                  function ()
                      local screen = mouse.screen
                      if tags[screen][i] then
                          awful.tag.viewtoggle(tags[screen][i])
                      end
                  end),
        awful.key({ modkey, "Shift" }, "#" .. i + 9,
                  function ()
                      if client.focus and tags[client.focus.screen][i] then
                          awful.client.movetotag(tags[client.focus.screen][i])
                      end
                  end),
        awful.key({ modkey, "Control", "Shift" }, "#" .. i + 9,
                  function ()
                      if client.focus and tags[client.focus.screen][i] then
                          awful.client.toggletag(tags[client.focus.screen][i])
                      end
                  end))
end

for i = 11, 22 do
    globalkeys = awful.util.table.join(globalkeys,
        awful.key({ modkey }, "F" .. i - 10,
                  function ()
                        local screen = mouse.screen
                        if tags[screen][i] then
                            awful.tag.viewonly(tags[screen][i])
			    awful.util.spawn(ffm)
                        end
                  end),
        awful.key({ modkey, "Control" }, "F" .. i - 10,
                  function ()
                      local screen = mouse.screen
                      if tags[screen][i] then
                          awful.tag.viewtoggle(tags[screen][i])
                      end
                  end),
        awful.key({ modkey, "Shift" }, "F" .. i - 10,
                  function ()
                      if client.focus and tags[client.focus.screen][i] then
                          awful.client.movetotag(tags[client.focus.screen][i])
                      end
                  end),
        awful.key({ modkey, "Control", "Shift" }, "F" .. i - 10,
                  function ()
                      if client.focus and tags[client.focus.screen][i] then
                          awful.client.toggletag(tags[client.focus.screen][i])
                      end
                  end))
 end

clientbuttons = awful.util.table.join(
    awful.button({ }, 1, function (c) client.focus = c; c:raise() end),
    awful.button({ modkey }, 1, awful.mouse.client.move),
    awful.button({ modkey }, 3, awful.mouse.client.resize))

-- Set keys
root.keys(globalkeys)
-- }}}

-- {{{ Rules
awful.rules.rules = {
    -- All clients will match this rule.
    { rule = { },
      properties = { border_width = beautiful.border_width,
                     border_color = beautiful.border_normal,
                     focus = true,
                     keys = clientkeys,
                     buttons = clientbuttons } },
    { rule = { class = "MPlayer" },
      properties = { floating = true } },
    { rule = { class = "pinentry" },
      properties = { floating = true } },
    { rule = { class = "gimp" },
      properties = { floating = true } },
    { rule = { class = "Vigor" },
      properties = { floating = true } },
    -- Set Firefox to always map on tags number 2 of screen 1.
    { rule = { class = "Firefox" },
      properties = { tag = tags[1][2] } },
    { rule = { class = "Conkeror" },
      properties = { tag = tags[1][2] } },
    { rule = { class = "Liferea" },
      properties = { tag = tags[1][3] } },
--    { rule = { class = "Emacs" },
--      properties = { tag = tags[1][4] } },
}
-- }}}

-- {{{ Signals
-- Signal function to execute when a new client appears.
client.add_signal("manage", function (c, startup)
    -- Add a titlebar
    -- awful.titlebar.add(c, { modkey = modkey })

    -- Enable sloppy focus
    c:add_signal("mouse::enter", function(c)
        if awful.layout.get(c.screen) ~= awful.layout.suit.magnifier
            and awful.client.focus.filter(c) then
            client.focus = c
        end
    end)

    if not startup then
        -- Set the windows at the slave,
        -- i.e. put it at the end of others instead of setting it master.
        -- awful.client.setslave(c)

        -- Put windows in a smart way, only if they does not set an initial position.
        if not c.size_hints.user_position and not c.size_hints.program_position then
            awful.placement.no_overlap(c)
            awful.placement.no_offscreen(c)
        end
    end
end)

client.add_signal("focus", function(c)
      c.border_color = beautiful.border_focus
      c.opacity = 1
end)
client.add_signal("unfocus", function(c)
      c.border_color = beautiful.border_normal
      c.opacity = 0.8
end)

-- client.add_signal("focus", function(c) c.border_color = beautiful.border_focus end)
-- client.add_signal("unfocus", function(c) c.border_color = beautiful.border_normal end)
-- }}}
